File: ios_app/iphone/ui/Tasks/Tab/TasksTabDropItem.swift
import SwiftUI
import shared

class TasksTabDropItem: ObservableObject {

    let name: String
    let square: Square

    init(name: String, square: Square) {
        self.name = name
        self.square = square
    }

    class Square {

        var x1: CGFloat
        var y1: CGFloat
        var x2: CGFloat
        var y2: CGFloat

        init(x1: CGFloat = 0, y1: CGFloat = 0, x2: CGFloat = 0, y2: CGFloat = 0) {
            self.x1 = x1
            self.y1 = y1
            self.x2 = x2
            self.y2 = y2
        }

        func upByRect(rect: CGRect) {
            x1 = rect.origin.x
            y1 = rect.origin.y
            x2 = rect.origin.x + rect.width
            y2 = rect.origin.y + rect.height
        }
    }
}

///

class TasksTabDropItemCalendar: TasksTabDropItem {

    init() {
        super.init(name: "Calendar", square: TasksTabDropItem.Square())
    }
}

class TasksTabDropItemTaskFolder: TasksTabDropItem {

    let taskFolderDb: TaskFolderDb

    init(_ taskFolderDb: TaskFolderDb) {
        self.taskFolderDb = taskFolderDb
        super.init(name: taskFolderDb.name, square: TasksTabDropItem.Square())
    }
}

File: android_app/src/main/java/me/timeto/app/ui/tasks/tab/TasksTabDropItem.kt
package me.timeto.app.ui.tasks.tab

import androidx.compose.ui.layout.LayoutCoordinates
import androidx.compose.ui.layout.positionInWindow
import me.timeto.shared.db.TaskFolderDb

sealed class TasksTabDropItem(
    val name: String,
    val square: Square,
) {

    fun upSquareByCoordinates(c: LayoutCoordinates) {
        val p = c.positionInWindow()
        square.x1 = p.x.toInt()
        square.y1 = p.y.toInt()
        square.x2 = p.x.toInt() + c.size.width
        square.y2 = p.y.toInt() + c.size.height
    }

    class Square(var x1: Int, var y1: Int, var x2: Int, var y2: Int)

    //
    // Types

    class Folder(
        val taskFolderDb: TaskFolderDb,
        square: Square,
    ) : TasksTabDropItem(taskFolderDb.name, square)

    class Calendar(
        square: Square,
    ) : TasksTabDropItem("Calendar", square)
}
